pkg/api/types.go

```diff
@@ -1410,6 +1412,55 @@ type ReplicationControllerList struct {
 	Items []ReplicationController `json:"items"`
 }
 
+// DisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type DisruptionBudget struct {
+	// Specification of the desired behavior of the DisruptionBudget.
+	Spec DisruptionBudgetSpec `json:"spec,omitempty"`
+	// Most recently observed status of the DisruptionBudget.
+	Status DisruptionBudgetStatus `json:"status,omitempty"`
+}
+
+// DisruptionBudgetSpec is a description of a DisruptionBudget.
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service.
+	// If there are multiple entries in the list, all must be satisfied for a disruption to be allowed.
+	RateLimits   []RateLimit `json:"rateLimits,omitempty"`
+
+	MinInstances MinInstances `json:"minInstances,omitempty"`
+}
+
+type MinInstances struct {
+	// The minimum number of pods that must be available simultaneously.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	MinAvailable intstr.IntOrString `json:"minAvailable"`
+	// Minimum number of seconds for which a newly created pod must be ready for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready).
+	MinReadySeconds int `json:"minReadySeconds"`
+}
+
+// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// Array of timestamps of recent disruptions corresponding to this DisruptionBudget.
+	// A timestamp is removed from the list when it is farther in the past than the maximum RateLimit.Duration.
+	DisruptionInfo []unversioned.Time `json:"disruptionInfo,omitempty"`
```

Comment:
etcd has a hard size limit, and with just one exception, which we're going to have to fix soon, I've tried to ensure we don't put unbounded-size lists into individual objects/keys.

An example use case might help.

* 1 disruption per 10 minutes
* 10 disruptions per day

I don't think we want to take the longest period, a day in this case, and divide it into N (say, 100) buckets, since that would lose too much accuracy for smaller periods. We probably need N buckets per time period, where N is on the order of 10 or 20. We should likely omit buckets with 0 events. We could identify the buckets by number (0 to N-1) in order to avoid repeating timestamps in each bucket. We could explicitly represent the number of buckets to allow us to change that in the future.

Something like:

An array of:
* period length
* number of buckets
* array of {bucket index, disruption count}

And one timestamp indicating the time of the last update of all the disruption counts.